### ISSUE TITLE
Show warnings for unused variables

### ISSUE THREAD

**Issue by aochagavia** [rust-analyzer/rust-analyzer#125]
Right now Rust Analyzer doesn't give much feedback while writing code. There is no autocompletion and you get only syntax errors. I was thinking about a good first step toward giving more feedback and came to the idea of showing warnings for unused variables. This should be doable by reusing `resolve_local_name`, which was defined in https://github.com/rust-analyzer/rust-analyzer/pull/98

In the future we will probably want to extend this to detect unused struct fields, types and other stuff. But we need to begin with *something*.

**Comment by aochagavia** [comment]
@matklad is it a problem to keep adding things while you are migrating everything to [salsa](https://github.com/salsa-rs/salsa)?

**Comment by aochagavia** [comment]
@matklad I would like to pick up this issue, but I am having trouble finding the right place to implement it. Do you have any suggestions? Tackling this means working at https://github.com/rust-analyzer/rust-analyzer/issues/27 as well, so a bit of mentoring would be greatly appreciated.

**Comment by matklad** [comment]
Oups, sorry, I've somehow missed all three comments on the issue :( 

I think that we already can detect unused fields, yeah. There might be false-positives due to macros, but they should be rate, and we can hack around them (var is used in a macro if macro's text contains var name), for example. 

The migration to salsa is mostly complete.

>Do you have any suggestions?

So, #27 is an unrelated issue: validators are purely for checking syntax (in general, we try to remove checks for the parser per se to make it more tolerant to the errors). 

I think the right place for this is near the `resolve_local_name`. So, something like

```
fn collect_unused_variables(
    fn_def: ast::FnDef
    scopes: &FnScopes
) -> Vec<ast::Name> {
}
```

should work.

**Comment by skade** [comment]
Is this issue still current?

![Peek 2020-04-14 17-38](https://user-images.githubusercontent.com/47542/79244197-e3203180-7e76-11ea-8bff-a9bad3e2a14d.gif)

Highlighting unused variables exists.

**Comment by lnicola** [comment]
It works by running `cargo check` and displaying its diagnostics, but it's not implemented natively in `rust-analyzer`. That might still happen, but probably not in the near future.

**Comment by matklad** [comment]
Ok, so this is handled by cargo check. Native handling is desirable, but then this issue is indistinguishable from "implement the rest of the ~~owl~~ compiler", so it's not useful.
